core/src/main/java/org/apache/iceberg/SnapshotScan.java

```diff
@@ -85,11 +85,18 @@ public ThisT useSnapshot(long scanSnapshotId) {
   }
 
   public ThisT useRef(String name) {
+    if (SnapshotRef.MAIN_BRANCH.equals(name)) {
+      return newRefinedScan(table(), tableSchema(), context());
+    }
+
     Preconditions.checkArgument(
         snapshotId() == null, "Cannot override ref, already set snapshot id=%s", snapshotId());
     Snapshot snapshot = table().snapshot(name);
     Preconditions.checkArgument(snapshot != null, "Cannot find ref %s", name);
-    return newRefinedScan(table(), tableSchema(), context().useSnapshotId(snapshot.snapshotId()));
+    return newRefinedScan(
+        table(),
+        SnapshotUtil.schemaFor(table(), name),
+        context().useSnapshotId(snapshot.snapshotId()));
```

Comment:
nit: Can we define an extra variable to so that this call will fit on one line?

```
TableScanContext newContext = context().useSnapshotId(snapshot.snapshotId());
return newRefinedScan(table(), SnapshotUtil.schemaFor(table(), name), newContext);
```